packages/go/cypher/model/pgsql/translate/tracking.go

```diff
@@ -0,0 +1,304 @@
+package translate
+
+import (
+	"fmt"
+	"github.com/specterops/bloodhound/cypher/model"
+	"github.com/specterops/bloodhound/cypher/model/pgsql"
+	"strconv"
+)
+
+type IdentifierGenerator map[pgsql.DataType]int
+
+func (s IdentifierGenerator) NewIdentifier(dataType pgsql.DataType) (pgsql.Identifier, error) {
+	var (
+		nextID    = s[dataType]
+		nextIDStr = strconv.Itoa(nextID)
+	)
+
+	// Increment the ID
+	s[dataType] = nextID + 1
+
+	switch dataType {
+	case pgsql.EdgeExpansion:
+		return pgsql.Identifier("ex" + nextIDStr), nil
+	case pgsql.PathComposite:
+		return pgsql.Identifier("p" + nextIDStr), nil
+	case pgsql.NodeComposite:
+		return pgsql.Identifier("n" + nextIDStr), nil
+	case pgsql.EdgeComposite:
+		return pgsql.Identifier("e" + nextIDStr), nil
+	default:
+		return "", fmt.Errorf("identifier with data type %s does not have a prefix case", dataType)
+	}
+}
+
+func NewIdentifierGenerator() IdentifierGenerator {
+	return IdentifierGenerator{}
+}
+
+type Constraint struct {
```

Comment:
trying to wrap my mind around these wonderful ancillary structures. What is a constraint?  where can a constraint appear in a cypher AST?  what gets stuffed into the dependencies of a constraint?  please leave some inline some comments in the code that illuminate some of these questions!